velox/experimental/wave/common/Atomic.cuh

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <breeze/platforms/platform.h>
+#include <breeze/utils/types.h>
+#include <breeze/platforms/cuda.cuh>
+#include "velox/experimental/wave/common/BitUtil.cuh"
+#include "velox/experimental/wave/common/CompilerDefines.h"
+
+namespace facebook::velox::wave {
+
+enum class MemoryScope { kDevice };
+
+enum class MemoryOrder {
```

Comment:
Do we have full barrier (something similar to MFENCE)?

Reply:
Not as part of this API at this time. You can use `__threadfence*` family of CUDA builtins directly but I'd be happy to add it to the breeze abstraction as a follow up if that makes sense.